devito/types/basic.py

```diff
@@ -550,12 +667,49 @@ def name(self):
     @property
     def indices(self):
         """The indices (aka dimensions) of the object."""
-        return self._indices
+        return self.args
+
+    @property
+    def index_ref(self):
+        """The indices (aka dimensions) of the object."""
+        return self._index_ref
+
+    @property
+    def ind_map(self):
+        return {d1: d2 for d1, d2 in zip(self.dimensions, self.index_ref)}
+
+    @property
+    def origin(self):
+        return tuple(r - d for r, d in zip(self.index_ref, self.dimensions))
 
     @property
     def dimensions(self):
         """Tuple of Dimensions representing the object indices."""
-        return self.indices
+        return self._dimensions
+
+    @property
+    def evaluate(self):
+        # Average values if at a location not on the Function's grid
+        weight = 1.0
+        avg_list = [self]
+        is_averaged = False
+        for i, ir, d in zip(self.indices, self.index_ref, self.dimensions):
+            off = (i - ir)/d.spacing
+            if not isinstance(off, sympy.Number) or int(off) == off:
+                pass
+            else:
+                weight *= 1/2
+                is_averaged = True
+                avg_list = [a.subs({i: i - d.spacing/2}) + a.subs({i: i + d.spacing/2})
+                            for a in avg_list]
+
+        if not is_averaged:
+            return self
+        return weight * sum(avg_list)
+
+    def index(self, dim):
```

Comment:
should this be a public method?

Reply:
I would think so

Reply:
why? I liked the use of `_at` in `_eval_at`. Can/should we do the same here? `_index_at` (subsuming "AT dimension ...")

Reply:
also, I don't like the placement of this method, in between all properties... could you move it to a better place